[SCM-981] Add run-its profile and fix broken tests

nielsbasjes opened the pull request:
Most of the tests are never run.
This patch:
- adds a tck-hg profile to run the HG specific tck tests.
- adds a run-its profile to run the all provider tck tests.
- Fixes the tests that fail. Those were all very small fixes so I include them here.

Now `mvn clean verify -Prun-its` runs all tck tests and passes (on my machine).

nielsbasjes commented:
Apparently the tests now pass (on Ubuntu) but because not all Github build environments/images have all the tools needed they now fail on (at least) the macOS image.
`Caused by: java.io.IOException: Cannot run program "hg" (in directory ...): error=2, No such file or directory`

Looking at https://github.com/apache/maven-gh-actions-shared/blob/v2/.github/workflows/maven-verify.yml I see that it simply uses the macOS-latest image provided by Github.

Please advise on how you would like to proceed.

michael-o commented:
There are tests which check wether `svn`, `git` as are installed via `assumeTrue`. Find those this you could apply to, I guess.

nielsbasjes commented:
> assumeTrue

Ah yes, interesting.
In the mean time I found an alternative way: Activate the tck-... profiles on a run-its property (not profile) in combination with the existence of the needed tool.

michael-o commented on maven-scm-providers/maven-scm-provider-hg/pom.xml:
```diff
@@ -66,10 +66,17 @@
+          <name>run-its</name>
+        </property>
+        <file>
+          <exists>/usr/bin/hg</exists>
```
Not portable

nielsbasjes replied:
I agree. I'll refactor using the assumeTrue you mentioned.

michael-o commented on maven-scm-providers/maven-scm-providers-git/pom.xml:
```diff
@@ -40,10 +40,18 @@
+          <name>run-its</name>
+        </property>
+        <file>
+          <exists>/usr/bin/git</exists>
```
Not portable

michael-o commented on maven-scm-providers/maven-scm-providers-svn/pom.xml:
```diff
@@ -43,6 +43,14 @@
+          <name>run-its</name>
+        </property>
+        <file>
+          <exists>/usr/bin/svn</exists>
```
Not portable

michael-o commented:
Please see: https://issues.apache.org/jira/browse/SCM-939

nielsbasjes commented:
Hmmm, this assumeTrue makes the test still fail eventhough the documentation indicates they should be skipped instead:
```
[ERROR] Errors: 
[ERROR]   JGitBranchCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitChangeLogCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitChangeLogCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitChangeLogCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitChangeLogCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitChangeLogCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitChangeLogCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitChangeLogCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitChangeLogCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitChangeLogCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitCheckInCommandCommitterAuthorTckTest.setUp:60->ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitCheckInCommandCommitterAuthorTckTest.setUp:60->ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitCheckInCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitCheckInCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitCheckOutCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitDiffCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitRemoteInfoCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitStatusCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitTagCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[ERROR]   JGitUntagCommandTckTest>ScmTckTestCase.setUp:136->ScmTckTestCase.checkScmPresence:124 » AssumptionViolated
[INFO] 
[ERROR] Tests run: 20, Failures: 0, Errors: 20, Skipped: 0
```

nielsbasjes commented:
Seems like the whole test suite is running on very old code.
When I add `@RunWith(JUnit4.class)` to ScmTckTestCase and add `@Test` on all the methods this suddenly starts working ... at the cost of failures in the `PlexusTestCase` area.

michael-o commented:
@cstamas is now working on replacing Plexus. Does this intersect?

michael-o commented:
@nielsbasjes Please rebase

nielsbasjes commented:
I have 
- Fixed a few broken tests.
- Added a tck-hg and tck-local for consistency
- Added a run-its profile that runs all.
- Added the method ScmTckTestCase::getScmProviderCommand which either returns null (i.e. no command needs to exist) or the name of the command that must exist.
This has been overridden in all `Hg...TckTest` , `SvnExe...TckTest` and `GitExe...TckTest` classes (except the two specials mentioned above). This method is checked before each test and ignores tests for which the required command is missing.

Classes with things I consider "unexpected" and have not changed:
- GitListCommandTckTest is abstract and not actually implemented anywhere.
- GitInfoCommandTckTest extends ScmTestCase instead of ScmTckTestCase

michael-o commented:
* `GitListCommandTckTest` hasn't obviously not being implemented...
* Looking at `GitInfoCommandTckTest` it does not seem seem to be a TCK test, but merely a unit test. It does not have the same structure as TCK tests.

michael-o commented on maven-scm-providers/maven-scm-provider-local/src/main/java/org/apache/maven/scm/provider/local/command/update/LocalUpdateCommand.java:
```diff
@@ -190,7 +190,7 @@ private List<ScmFile> update( File source, File baseDestination, List<File> file
             int chop = baseDestination.getAbsolutePath().length();
 
-            String fileName = "/" + destinationFile.getAbsolutePath().substring( chop + 1 );
+            String fileName = destinationFile.getAbsolutePath().substring( chop + 1 );
```
Interesting that blindly a slash was prepended. That was doomed to fail on Windows, no?

nielsbasjes replied:
I have Ubuntu and it fails because the assertion expects it without a slash.

nielsbasjes replied:
And yes, this indeed a test that would probably fail on Windows too.

michael-o commented on maven-scm-test/src/main/java/org/apache/maven/scm/tck/command/changelog/ChangeLogCommandTckTest.java:
```diff
@@ -94,7 +94,6 @@ public void testChangeLogCommand()
 
-
-        assertEquals( COMMIT_MSG, changeset.getComment() );
+        assertTrue( changeset.getComment().startsWith( COMMIT_MSG ) ); // Sometimes there is a newline appended
```
Under which circumstances did this occur to you?

nielsbasjes replied:
One of the SCMs returned it with a newline. Perhaps this should be fixed differently.

michael-o replied:
Which? I would rather strip trailing LS and then go through `assertEquals()`

nielsbasjes replied:
Turns out that the HG code ... appends a newline. Fixed

michael-o commented on maven-scm-test/src/main/java/org/apache/maven/scm/ScmTckTestCase.java:
```diff
@@ -112,14 +114,19 @@ protected List<String> getScmFileNames()
+                    scmProviderCommand, ScmTestCase.isSystemCmd( scmProviderCommand ) );
             assumeTrue( "Skipping tests because the required command " + scmProviderCommand + " is not available.",
                 ScmTestCase.isSystemCmd( scmProviderCommand ) );
+            LOG.warn( "Required SCM command \"{}\" exists.", scmProviderCommand );
```
Why warn? Is this intermediate?

nielsbasjes replied:
I added this just now to figure out why it still fails on macOS.

nielsbasjes replied:
I have just now uninstalled hg from my local system and I'm now able to locally reproduce the problem.

nielsbasjes replied:
Found it. Multiple `@Before` and no ordering guarantee among these.

michael-o commented on maven-scm-providers/maven-scm-provider-hg/src/test/java/org/apache/maven/scm/provider/hg/HgRepoUtils.java:
```diff
@@ -35,9 +35,11 @@
  * @author <a href="mailto:thurner.rupert@ymono.net">thurner rupert</a>
  */
-public class HgRepoUtils
+public abstract class HgRepoUtils
```
Why? This class even extends `ScmTestCase` for whatsoever reason and has no descendants.

nielsbasjes replied:
This failed in the previous testing framework. It no longer does so I have remove this change.

michael-o commented:
You can rename: GitInfoCommandTckTest

nielsbasjes commented:
Now all tests are run on all platforms where the needed tooling has been installed.
I put up this followup issue to improve the build environment to go towards all tests being run on all systems always.

https://issues.apache.org/jira/browse/SCM-988

michael-o commented on maven-scm-providers/maven-scm-provider-hg/src/main/java/org/apache/maven/scm/provider/hg/command/changelog/HgChangeLogConsumer.java:
```diff
@@ -148,7 +148,6 @@ else if ( line.startsWith( MESSAGE_TOKEN ) )
         {
             StringBuilder comment = new StringBuilder( currentChange.getComment() );
             comment.append( line );
-            comment.append( '\n' );
```
WTF? Who had this idea to modify the original comment?

nielsbasjes replied:
Removing this seemed the cleanest solution to me.

michael-o commented:
Working on this...
```
[INFO]
[ERROR] Failures:
[ERROR]   HgDiffCommandTckTest.testDiffCommand:86->ScmTckTestCase.addToWorkingTree:218 Expected 1 file in the added files list [] expected:<1> but was:<0>
[ERROR]   HgStatusCommandTckTest>StatusCommandTckTest.testStatusCommand:107->ScmTckTestCase.addToWorkingTree:218 Expected 1 file in the added files list [] expected:<1> but was:<0>
[ERROR]   HgUpdateCommandTckTest>UpdateCommandTckTest.testUpdateCommand:119->ScmTckTestCase.addToWorkingTree:218 Expected 1 file in the added files list [] expected:<1> but was:<0>
[INFO]
[ERROR] Tests run: 17, Failures: 3, Errors: 0, Skipped: 0
```

with
```
$ hg --version
Mercurial Distributed SCM (version 6.1.1)
(see https://mercurial-scm.org for more information)

Copyright (C) 2005-2022 Olivia Mackall and others
This is free software; see the source for copying conditions. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```
on
```
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
mosipov@bsd1srv:/usr/home/mosipov/var/Projekte/maven-scm (SCM-981 =)
$ mvn -v
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /usr/local/apache-maven-3.6.3
Java version: 1.8.0_332, vendor: OpenJDK BSD Porting Team, runtime: /usr/local/openjdk8/jre
Default locale: de_DE, platform encoding: UTF-8
OS name: "freebsd", version: "12.3-stable", arch: "amd64", family: "unix"
```

michael-o reviewed (commented):
Thanks for the work!

jira-importer commented:
Resolve #1211
